### ISSUE TITLE
Switch from public-inbox.org to lore.kernel.org

### ISSUE THREAD

**Issue by dscho** [gitgitgadget/gitgitgadget#155]
Seems that public-inbox.org will go the same way as gmane.org went, or almost: a new archive has been set up at lore.kernel.org. See https://lore.kernel.org/git/20191120195556.GA25189@dcvr/t/#u (which is identical to https://public-inbox.org/git/20191120195556.GA25189@dcvr/t/#u, and has the message by the public-inbox.org maintainer).

**Comment by webstech** [comment]
In _lookup-commit.sh_, should this be changed to somewhere else?
`git clone --bare https://git-for-windows.visualstudio.com/git.public-inbox/_git/git.public-inbox "$PUBLIC_INBOX_DIR"`

**Comment by webstech** [comment]
I have a [branch](https://github.com/webstech/gitgitgadget/compare/mail-archive) for this if you are interested.  Not sure how soon you want to get this done.  Pending the previous comment.  Also, need to run some tests and verify this:
`            this.state.latestRevision =
                "cf3590b3a1ce08a52b01142307b8fcc089acb6a6";
`

**Comment by dscho** [comment]
>  Also, need to run some tests and verify this:
> ` this.state.latestRevision = "cf3590b3a1ce08a52b01142307b8fcc089acb6a6";`

I just tried to clone from lore.kernel.org, no dice. So it is definitely not a full replacement yet.

> I have a [branch](https://github.com/webstech/gitgitgadget/compare/mail-archive) for this if you are interested.

Nice!

About https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archive#diff-cf3021f43051c9b73cb47ee8437f78faL87-R92: I think we need to fall back to looking for `public-inbox.org` links if `lore.kernel.org` was not a match.

> Not sure how soon you want to get this done.

Well, seems that efforts are under way to move git/git to using the new URL, see https://lore.kernel.org/git/20191127200421.GA27456@dcvr/T/#u and https://lore.kernel.org/git/14167012a13125fe3c31af27afe68f84c9cdd2a2.1574882812.git.liu.denton@gmail.com/T/#u

So I'd say as soon as they have a working Git repository to clone from, we should switch.

**Comment by dscho** [comment]
Seems like https://lore.kernel.org/git/0 is the URL for the repository. But the commit hashes do not match... I guess lore.kernel.org set up their own email account to receive those mails from...

```sh
$ git ls-remote https://lore.kernel.org/git/0 master
5d15b4faaaff41fbd9d53e392014254294ae7287        refs/heads/master

$ git ls-remote https://public-inbox.org/git master
52f48398dda94ce89b71c004eccd817c34c1e1a3        refs/heads/master
```

**Comment by dscho** [comment]
> In _lookup-commit.sh_, should this be changed to somewhere else?
> `git clone --bare https://git-for-windows.visualstudio.com/git.public-inbox/_git/git.public-inbox "$PUBLIC_INBOX_DIR"`

You're right, I will need to set up a mirror for that, together with an Azure Pipeline to keep it in sync.

**Comment by dscho** [comment]
> I will need to set up a mirror for that, together with an Azure Pipeline to keep it in sync.

https://dev.azure.com/gitgitgadget/git/_git/lore-git?_a=history and https://dev.azure.com/gitgitgadget/git/_build?definitionId=14&_a=summary.

**Comment by webstech** [comment]
> > In _lookup-commit.sh_, should this be changed to somewhere else?
> > `git clone --bare https://git-for-windows.visualstudio.com/git.public-inbox/_git/git.public-inbox "$PUBLIC_INBOX_DIR"`
> 
> You're right, I will need to set up a mirror for that, together with an Azure Pipeline to keep it in sync.

1. Is lookup-commit.sh used?
2. If 1=yes should the git clone point to the new azure repo or the lore.kernel repo instead of visualstudion.com?
3. Do you want me to toss in corrections to `PUBLIX_INBOX_DIR` typos?

**Comment by webstech** [comment]
> Seems like https://lore.kernel.org/git/0 is the URL for the repository. But the commit hashes do not match... I guess lore.kernel.org set up their own email account to receive those mails from...

Can't wait for the next epoch.  😲 

Am guessing the `state.latestRevision` in the notes will need to be updated manually at the time this change goes in?

**Comment by webstech** [comment]
> About [master...webstech:mail-archivediff-cf3021f43051c9b73cb47ee8437f78faL87-R92](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archivediff-cf3021f43051c9b73cb47ee8437f78faL87-R92): I think we need to fall back to looking for `public-inbox.org` links if `lore.kernel.org` was not a match.

That link is a no go but it appears to be line 92 in patch-series.ts where I did add the fall back check.  Am I misunderstanding your comment?  Is there somewhere else a fall back check is needed?  I see a check for gmane is still there.

> So I'd say as soon as they have a working Git repository to clone from, we should switch.

I pushed an updated to lookup-commit.sh to fetch from the correct repo.  That appears to be the only place that was needed.

Config will need to be updated `gitgitgadget.publicInboxGitDir` when this goes in.

**Comment by dscho** [comment]
> * Is lookup-commit.sh used?

Yes, in the [Update GitGitGadget's commit to mail notes](https://dev.azure.com/gitgitgadget/git/_build/index?definitionId=9) Azure Pipeline.

Granted, I don't think that we rely on `refs/notes/mail-to-commit` any longer, but instead use `git range-diff upstream/pu...` to identify upstream commits (which is *a ton* more reliable).

Still, I would love to have this updated, too.

> * If 1=yes should the git clone point to the new azure repo or the lore.kernel repo instead of visualstudion.com?

I would like to stay with the Azure Repos (https://dev.azure.com/git-for-windows is the new way to reference https://git-for-windows.visualstudio.com/; Azure DevOps really is independent of Visual Studio, hence that change) rather than `lore.kernel.org`: the Azure Pipeline's build agents are a lot closer to Azure Repos, so any network operation using Azure Repos is faster and more robust.

> * Do you want me to toss in corrections to `PUBLIX_INBOX_DIR` typos?

Oy. Yes, please 🤦‍♂ 

> > Seems like https://lore.kernel.org/git/0 is the URL for the repository. But the commit hashes do not match... I guess lore.kernel.org set up their own email account to receive those mails from...
> 
> Can't wait for the next epoch. 😲

Yes :-(

> Am guessing the `state.latestRevision` in the notes will need to be updated manually at the time this change goes in?

Yes, unfortunately: https://dev.azure.com/gitgitgadget/git/_git/lore-git/commit/cf3590b3a1ce08a52b01142307b8fcc089acb6a6/ says:

```
The given commit was not found.
Return to the repository history
```

I believe I found the correct revision: https://dev.azure.com/gitgitgadget/git/_git/lore-git/commit/3b38d8d206c64bf3dc873ba8ae9dbd48ed43f612/ has the same `Message-ID` as https://dev.azure.com/git-for-windows/git.public-inbox/_git/git.public-inbox/commit/cf3590b3a1ce08a52b01142307b8fcc089acb6a6/: 20180616191400.GB8914@sigill.intra.peff.net.

Therefore, I think cf3590b3a1ce08a52b01142307b8fcc089acb6a6 needs to be replaced with 3b38d8d206c64bf3dc873ba8ae9dbd48ed43f612 when switching to lore.kernel.org/git.

> > About [master...webstech:mail-archivediff-cf3021f43051c9b73cb47ee8437f78faL87-R92](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archivediff-cf3021f43051c9b73cb47ee8437f78faL87-R92): I think we need to fall back to looking for `public-inbox.org` links if `lore.kernel.org` was not a match.
> 
> That link is a no go

It was broken due to the colon, which apparently is not handled well by GitHub's Markdown processor. What I meant is [this link](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archive#diff-cf3021f43051c9b73cb47ee8437f78faL87-R92).

> where I did add the fall back check.

You [totally did](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archive#diff-cf3021f43051c9b73cb47ee8437f78faR89-R92). My bad for overlooking it, somehow my visual cortex already equates "public-inbox.org" with "lore.kernel.org".

> Config will need to be updated `gitgitgadget.publicInboxGitDir` when this goes in.

I would like to use a new config setting there, say, `gitgitgadget.loreGitDir`? Likewise, I would like to replace the name `PUBLIC_INBOX_DIR` with `LORE_GIT_DIR`, just so that I can catch mistakes I make on the Azure Pipelines side early.

In the same spirit, how would you feel about renaming `PublicInboxGitHelper` to `LoreGitHelper`? And `publicInboxGitDir` to `loreGitDir`?

BTW I think that [this change](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archive#diff-ba4575a602122b5634a18cff2159791eR55) should look more like this:

```sh
		echo "$m" | sed -e 's|^|https://public-inbox.org/git/|' -e 's|^|https://lore.kernel.org/git/|'
```

**Comment by dscho** [comment]
> BTW I think that [this change](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archive#diff-ba4575a602122b5634a18cff2159791eR55) should look more like this:
> 
> ```shell
> 		echo "$m" | sed -e 's|^|https://public-inbox.org/git/|' -e 's|^|https://lore.kernel.org/git/|'
> ```

Whoooooops. This suggestion is incorrect.

Instead, I would suggest to _replace_ [the line](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archive#diff-ba4575a602122b5634a18cff2159791eR54):

```sh
echo "$m" | sed 's|^|https://public-inbox.org/git/|'
```

by [this line](https://github.com/gitgitgadget/gitgitgadget/compare/master...webstech:mail-archive#diff-ba4575a602122b5634a18cff2159791eR55):

```sh
echo "$m" | sed 's|^|https://lore.kernel.org/git/|'
```

(instead of just adding the latter).

**Comment by webstech** [comment]
> I would like to stay with the Azure Repos (https://dev.azure.com/git-for-windows is the new way to reference https://git-for-windows.visualstudio.com/; Azure DevOps really is independent of Visual Studio, hence that change) rather than `lore.kernel.org`: the Azure Pipeline's build agents are a lot closer to Azure Repos, so any network operation using Azure Repos is faster and more robust.

Is this the new URL?  (no git-for-windows)
`git clone --bare https://dev.azure.com/gitgitgadget/git/_git/lore-git`

> > Am guessing the `state.latestRevision` in the notes will need to be updated manually at the time this change goes in?

...
> Therefore, I think cf3590b3a1ce08a52b01142307b8fcc089acb6a6 needs to be replaced with 3b38d8d206c64bf3dc873ba8ae9dbd48ed43f612 when switching to lore.kernel.org/git.
 
I have updated in the code but that is the start if no state is found.  The `state: IGitMailingListMirrorState` in the notes in the repo will need to change to match the revision in the lore.get repo that matches the revision in public-inbox at the time of implementation.

> I would like to use a new config setting there, say, `gitgitgadget.loreGitDir`? Likewise, I would like to replace the name `PUBLIC_INBOX_DIR` with `LORE_GIT_DIR`, just so that I can catch mistakes I make on the Azure Pipelines side early.
> 
> In the same spirit, how would you feel about renaming `PublicInboxGitHelper` to `LoreGitHelper`? And `publicInboxGitDir` to `loreGitDir`?

Checking config for gitgitgadget.loreGitDir
Changed PUBLIC_INBOX_DIR to LORE_GIT_DIR
Took the liberty of being less specific
Renamed PublicInboxGitHelper -> MailArchiveGitHelper (and file renamed)
Renamed file search-commit-on-public-inbox.sh -> search-commit-on-mail-archive.sh
Renamed publicInboxGitDir -> mailArchiveGitDir

**Comment by dscho** [comment]
> Is this the new URL? (no git-for-windows)
> `git clone --bare https://dev.azure.com/gitgitgadget/git/_git/lore-git`

Yes. IIRC I added the public-inbox mirror long time before GitGitGadget was born. But technically, the mirror is relevant not to the Git for Windows project, but to the GitGitGadget project.

> > Therefore, I think cf3590b3a1ce08a52b01142307b8fcc089acb6a6 needs to be replaced with 3b38d8d206c64bf3dc873ba8ae9dbd48ed43f612 when switching to lore.kernel.org/git.
> 
> I have updated in the code but that is the start if no state is found. The `state: IGitMailingListMirrorState` in the notes in the repo will need to change to match the revision in the lore.get repo that matches the revision in public-inbox at the time of implementation.

True. That might be a bit tricky. The cleanest way would probably be to provide some special code intended solely for the transition, something like this:

```diff
diff --git a/lib/mail-archive-helper.ts b/lib/mail-archive-helper.ts
index 9685d55..35ab825 100644
--- a/lib/mail-archive-helper.ts
+++ b/lib/mail-archive-helper.ts
@@ -180,7 +180,43 @@ export class MailArchiveGitHelper {
              */
             this.state.latestRevision =
                 "3b38d8d206c64bf3dc873ba8ae9dbd48ed43f612";
+        } else if (await revParse(this.state.latestRevision,
+                                  this.mailArchiveGitDir) === undefined) {
+            const publicInboxGitDir = process.env.PUBLIC_INBOX_DIR;
+            if (publicInboxGitDir === undefined) {
+                throw new Error(`Commit ${this.state.latestRevision
+                                } not found; need PUBLIC_INBOX_DIR`);
+            }
+            let commitDiff =
+                await git(["show", this.state.latestRevision, "--"],
+                          { workDir: publicInboxGitDir });
+            if (!commitDiff) {
+                throw new Error(`Could not find ${this.state.latestRevision
+                                } in ${publicInboxGitDir}`);
+            }
+            const match = commitDiff.match(/\n\+(Message-ID: [^\n]*)/);
+            if (!match) {
+                throw new Error(`Could not find Message-ID in ${commitDiff}`);
+            }
+            let commit = "HEAD";
+            for (;;) {
+                commit = await git(["log", "-1", "--format=%H", `-S${match[1]}`,
+                                    commit, "--"],
+                                   { workDir: this.mailArchiveGitDir});
+                if (!commit) {
+                    throw new Error(`Could not find ${this.state.latestRevision
+                                }'s equivalent in ${this.mailArchiveGitDir}`);
+                }
+                commitDiff = await git(["show", commit, "--"],
+                                       {workDir: this.mailArchiveGitDir});
+                if (commitDiff.indexOf(`+${match[1]}`) >= 0) {
+                    break;
+                }
+                commit += "^"; /* continue search with the parent commit */
+            }
+            this.state.latestRevision = commit;
         }
+
         const head = await revParse("master", this.mailArchiveGitDir);
         if (this.state.latestRevision === head) {
             return false;
```

Once we're safely past merging the PR with your changes, we can easily kick this out later. For your convenience, I just pushed this to https://github.com/dscho/gitgitgadget/commits/mail-archive

> Checking config for gitgitgadget.loreGitDir
> Changed PUBLIC_INBOX_DIR to LORE_GIT_DIR
> Took the liberty of being less specific
> Renamed PublicInboxGitHelper -> MailArchiveGitHelper (and file renamed)
> Renamed file search-commit-on-public-inbox.sh -> search-commit-on-mail-archive.sh
> Renamed publicInboxGitDir -> mailArchiveGitDir

Nice!

**Comment by dscho** [comment]
> I just pushed this to https://github.com/dscho/gitgitgadget/commits/mail-archive

I just realized that just looking for that `+${match[1]}` won't be enough: somebody could have mentioned the full message ID, together with the "Message-ID: " prefix, in a mail. I just force-pushed an update that requires the search needle to be surrounded by newlines.

Technically, this is _still_ not correct, as we would need to verify that the empty line separating the mail header from its body (if any such line is found) needs to come _after_ the found header. But I think at some stage I need to stop refining this code that will be needed for one single occasion only, anyway.

**Comment by webstech** [comment]
> Technically, this is _still_ not correct, as we would need to verify that the empty line separating the mail header from its body (if any such line is found) needs to come _after_ the found header. But I think at some stage I need to stop refining this code that will be needed for one single occasion only, anyway.

Maybe the `commitDiff` set at line 190 could have been compared with the one obtained from the new repo (ie the commit messages are the same).  But, as you said it is a one time use thing (until the epoch shows up).

Now that you have everything, do you want to do the pull request?

**Comment by dscho** [comment]
> 
> 
> > Technically, this is _still_ not correct, as we would need to verify that the empty line separating the mail header from its body (if any such line is found) needs to come _after_ the found header. But I think at some stage I need to stop refining this code that will be needed for one single occasion only, anyway.
> 
> Maybe the `commitDiff` set at line 190 could have been compared with the one obtained from the new repo (ie the commit messages are the same).

Sadly, no. The problem is that the v1 format (which is used by https://public-inbox.org/git/) _adds_ the mails to a humongous worktree. The v2 format (used by https://lore.kernel.org/git/0) instead replaces the contents of the file `m`. That is, the publicInbox commitDiff will not remove any lines, ever, the loreGit commitDiff will replace the entire contents and therefore remove all the lines (in practice, there are often empty lines that are mistaken for being matching context lines).

> But, as you said it is a one time use thing (until the epoch shows up).

I don't actually think that this will a problem when epoch shows up, as the commit history between epochs is not overlapping (as far as I understand).

For epoch transitions, we will instead need an Azure Pipeline that regularly pings the next epoch repository (or a file, Eric Wong, the inventor of the code backing both mail archives, [ruminated](https://lore.kernel.org/git/20191128194829.GA15521@dcvr/) about adding the relevant information to https://lore.kernel.org/git/_/text/config/raw). In my mind, this Azure Pipeline will commit that epoch change to GitGitGadget and open a PR, _and_ adjust the build variables referencing the repository URL in the relevant Azure pipelines. That way, the next fetch will fetch "unrelated" commit history, force-update the `master` branch, but still find the old revision.

And `git log ${latestRevision}..master` will be equivalent to `git log master` because `${latestRevision}` will simply be unreachable from `master`, and that is exactly what we would want in that case.

Or maybe it should be a more manual process, I don't know.

> Now that you have everything, do you want to do the pull request?

Yes! Thank you for working on this, much appreciated!

**Comment by webstech** [comment]
> > Now that you have everything, do you want to do the pull request?
> 
> Yes! Thank you for working on this, much appreciated!

Not sure if you got busy but just wanted to make sure it was clear I meant you would do the pull request since you have both commits in your branch.

**Comment by dscho** [comment]
Ooops. I indeed thought that you would like to cherry-pick my commit and open the PR. I will open it tomorrow.

**Comment by dscho** [comment]
Fixed via #164
